### PR TITLE
feat(engine-v2): Add support for SSE subscriptions to subgraphs.

### PR DESCRIPTION
**Pull request by Finistere**
If the websocket URL is not defined we'll send a request with `Accept:
text/event-stream,application/json;q=0.9` to start a SSE subscriptions.
We may want to allow a different URL for this at some point, not sure.

I'm using the quick start of the JS `graphql-sse` library as a SSE
subgraph in the integration-tests.

I've changed the `Fetcher` trait to remove a bunch of allocations we're
doing for websockets, couldn't resist not removing those.

### CODE REVIEW

File: engine/crates/engine-v2/src/engine/retry_budget.rs [hunk 1]
```diff
@@ -33,7 +33,7 @@ impl RetryBudgets {
 }
 
 impl<R: Runtime> super::Engine<R> {
-    pub(crate) fn get_retry_budget_for_query(&self, endpoint_id: GraphqlEndpointId) -> Option<&Budget> {
+    pub(crate) fn get_retry_budget_for_non_mutation(&self, endpoint_id: GraphqlEndpointId) -> Option<&Budget> {
```
Comment (tomhoule) [inline engine/crates/engine-v2/src/engine/retry_budget.rs:36]: :+1:

File: engine/crates/integration-tests/data/sse-subgraph/tmp.html [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+   *
+   * This source code is licensed under the MIT license found in the
+   * LICENSE file in the root directory of this source tree.
+   *)
+*/
+;
+      ;globalThis[Symbol.for("Bun__renderFallbackError")](globalThis.__BUN_DATA__);
+    </script>
+  </body>
+</html>
```
Comment (tomhoule) [inline engine/crates/integration-tests/data/sse-subgraph/tmp.html:113]: Is that committed intentionally?

File: engine/crates/integration-tests/data/tmp.html [hunk 3]
```diff
@@ -0,0 +1 @@
+{"data":{"hello":"world"}}
```
Comment (tomhoule) [inline engine/crates/integration-tests/data/tmp.html:1]: (same here)

File: engine/crates/runtime/src/fetch.rs [hunk 4]
```diff
@@ -20,53 +21,135 @@ impl FetchError {
 
 pub type FetchResult<T> = Result<T, FetchError>;
 
-// very minimal for now, but will be expanded as we need it.
-pub struct FetchRequest<'a> {
-    pub url: &'a url::Url,
+/// reqwest uses Url instead of Uri, so as long as it's the actual implementation underneath it's a
+/// bit of a waste to use http::Request
```
Comment (tomhoule) [inline engine/crates/runtime/src/fetch.rs:25]: I don't understand this comment — what's URI in this context?

**Review by tomhoule — APPROVED** [review]

File: engine/crates/integration-tests/data/sse-subgraph/tmp.html (hunk 2 above)
Comment (Finistere) [reply engine/crates/integration-tests/data/sse-subgraph/tmp.html:113]: oops, thanks

File: engine/crates/runtime/src/fetch.rs (hunk 4 above)
Comment (Finistere) [reply engine/crates/runtime/src/fetch.rs:25]: meant this: https://docs.rs/http/latest/http/uri/struct.Uri.html
instead of `url::Url`

Comment (tomhoule) [reply engine/crates/runtime/src/fetch.rs:25]: Aah ok, makes sense

**Pull request merged**